lib/diamond/utils.ts

```diff
@@ -0,0 +1,305 @@
+      ) {
+        facetCuts.push(
+          printFacetCuts(AddressZero, [selector], FacetCutAction.REMOVE),
+        );
+
+        selectorsRemoved = true;
+      }
+    }
+  }
+
+  if (!selectorsRemoved) {
+    throw new Error('No selectors were removed from FacetCut');
+  }
+
+  return groupFacetCuts(facetCuts);
+}
+
+// preview a FacetCut which adds, replaces, or removes selectors, as needed
+export async function previewFacetCut(
+  diamond: IDiamondReadable,
+  contracts: Contract[],
+  only: FacetFilter[][] = [[], [], []],
+  except: FacetFilter[][] = [[], [], []],
+): Promise<FacetCut[]> {
+  let addFacetCuts: FacetCut[] = [];
+  let replaceFacetCuts: FacetCut[] = [];
+  let removeFacetCuts: FacetCut[] = [];
+
+  try {
+    addFacetCuts = await addUnregisteredSelectors(
+      diamond,
+      contracts,
+      only[0],
+      except[0],
+    );
+  } catch (error) {
+    console.log(`WARNING: ${(error as Error).message}`);
+  }
+
+  try {
+    replaceFacetCuts = await replaceRegisteredSelectors(
+      diamond,
+      contracts,
+      only[1],
+      except[1],
+    );
+  } catch (error) {
+    console.log(`WARNING: ${(error as Error).message}`);
+  }
+
+  try {
+    removeFacetCuts = await removeRegisteredSelectors(
+      diamond,
+      contracts,
+      only[2],
+      except[2],
+    );
+  } catch (error) {
+    console.log(`WARNING: ${(error as Error).message}`);
+  }
+
+  return groupFacetCuts([
+    ...addFacetCuts,
+    ...replaceFacetCuts,
+    ...removeFacetCuts,
+  ]);
+}
+
+// executes a DiamondCut using the provided FacetCut
+export async function diamondCut(
+  diamond: IDiamondWritable,
+  facetCut: FacetCut[],
+  target: string = AddressZero,
+  data: string = '0x',
+) {
+  (await diamond.diamondCut(facetCut, target, data)).wait(1);
+}
+
+// groups facet cuts by target address and action type
+export function groupFacetCuts(facetCuts: FacetCut[]): FacetCut[] {
+  const cuts = facetCuts.reduce((acc: FacetCut[], facetCut: FacetCut) => {
+    if (acc.length == 0) acc.push(facetCut);
+
+    let exists = false;
+
+    acc.forEach((_, i) => {
+      if (
+        acc[i].action == facetCut.action &&
+        acc[i].target == facetCut.target
+      ) {
+        acc[i].selectors.push(...facetCut.selectors);
+        // removes duplicates, if there are any
+        acc[i].selectors = [...new Set(acc[i].selectors)];
+        exists = true;
+      }
+    });
+
+    // push facet cut if it does not already exist
+    if (!exists) acc.push(facetCut);
+
+    return acc;
+  }, []);
+
+  let cache: any = {};
+
+  // checks if selector is used multiple times, emits warning
+  cuts.forEach((cut) => {
+    cut.selectors.forEach((selector: string) => {
+      if (cache[selector]) {
+        console.log(
+          `WARNING: selector: ${selector}, target: ${cut.target} is defined in multiple cuts`,
+        );
+      } else {
+        cache[selector] = true;
+      }
+    });
+  });
+
+  return cuts;
+}
+
+export function printFacetCuts(
```

Comment:
This doesn't seem to print anything.

Reply:
https://github.com/solidstate-network/solidstate-solidity/pull/158/commits/f4c211b7883346956d256edcaac04fa951251fdc